third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -1901,9 +1901,13 @@ impl ExpTranslator<'_, '_, '_> {
                 } else {
                     self.translate_exp(exp, &target_ty)
                 };
-                if self.subs.specialize(&target_ty).is_reference() {
+                let specialized_target_ty = self.subs.specialize(&target_ty);
+                if specialized_target_ty.is_reference() {
                     self.error(&loc, "cannot borrow from a reference")
                 }
+                if specialized_target_ty.is_tuple() {
+                    self.error(&loc, "cannot borrow a tuple")
```

Comment:
We are actually allowing `let (_x, _y) = *&(a, b);` and let x = &(a)`, could you double check what happens after your fix?

Reply:
For `let x = &(a);`: this is allowed. Note that `(a)` is not parsed as a tuple. I can add this in as a test case.
For `let (_x, _y) = *&(a, b);`; this is not allowed with my fix. And I see that this is allowed with current move compiler. Is this intentional though or bug? The move book states "The only operation that can be done on tuples currently is destructuring." and this seems to break that logic.

Reply:
yeah, as far as I remember, this may relate to optimizing `*&` away, which is necessary for testing variants of enum during generation of stackless bytecode. Any insights on whether we should allow this, @vineethk?

Reply:
Yeah it has to be due to optimizing. As another data point, the current compiler will reject this
```
fun g() : &(u64, u64) {
    return &(1, 2)
}
fun f() {
    let (x, y) = *g();
}
```

Despite accepting `let (x, y) = *&(1, 2);`, which is a bit counter-intuitive.

Reply:
I think it is okay to reject `*&(a, b)`: this may just have an oversight when the optimization was implemented.

The fix looks reasonable, can we please add more tests? Some thoughts:
- `&foo()` where `foo` is an inline function returning a tuple
- same as above, but `foo` is a regular function
- take `&mut` of a tuple and try to assign into it
- some of the examples discussed above.

Can approve after more tests are added.

Reply:
The following test case:
```
fun ret_borrow_tup(): &(u64, u64) {
    abort 0
}
```
seems to generate a compiler ICE.

This is certainly a bug. You can look into fixing it in this PR (as it is closely related), or create another PR for it.

Reply:
I'll look into this in separate PR; going to spend rest of today addressing your comments to my other outstanding PR + focusing on Etna.